### ISSUE TITLE
Possible performance issue in docker build step

### ISSUE THREAD

**Issue by swiedenfeld** [docker-java/docker-java#1409]
Environment: IntelliJ IDEA 2020.2 EAP Build #IU-202.5103.13 with bundled Docker plugin 202.5103.13. Docker plugin must use docker-java, connection to Docker Desktop for Windows via npipe. Cannot find out the exact version of docker-java that is used.

Given the following minimal project setup:
`node_modules` with +67k files
`Dockerfile` with single line `FROM hello-world` 
`.dockerignore` with single line `node_modules`

Then, run `docker build .` in the root directory.

Expected behaviour: The image is built instantly, because build context contains only the Dockerfile (large node_modules directory is ignored), and there are almost no instructions in the Dockerfile itself.

Observed behaviour: For +30 seconds, the screen shows "Building image..." , and after that only one file is added to the build context archive.

I assume, the build traverses all +67k files in the node_modules directory, only to find none of them have to be included. This would be very inefficient. When there are no exceptions, that allow a file below node_modules, it might skip the whole directory.

Update:
Unit test in `docker-java-core`, `testdir` has the large folder as child, `.dockerignore` absent:
```java
@Test
    public void can_parse_dockerfile() throws IOException {
        File dockerFile = new File(".\\testdir\\Dockerfile");
        File baseDirectory = new File(".\\testdir");
        Dockerfile dockerfile = new Dockerfile(dockerFile, baseDirectory);
        Stopwatch stopwatch = Stopwatch.createStarted();
        Dockerfile.ScannedResult parse = dockerfile.parse();
        System.out.println("Parse took " + stopwatch.elapsed(TimeUnit.MILLISECONDS) + "ms");
        System.out.println(parse.filesToAdd.size() + " files added");
    }
```
Output:
> Parse took 28008ms
> 4 files added

**Comment by KostyaSha** [comment]
Try `RUN ls` if there are files then they were sent to build context into daemon. 


> Cannot find out the exact version of docker-java that is used.

Please fill issue against IDEA first, dockefile parsing was updated few times in different versions.

**Comment by viatra2** [comment]
IDEA plugin overrides the `com.github.dockerjava.core.dockerfile.Dockerfile` in order to provide some feedback in the build log during potentially long archive creation. 

However, the reported issue comes from the recursive call `ScannedResult # addFilesInDirectory` [here](https://github.com/docker-java/docker-java/blob/e1586ccb94140d6c3c2cfd28d50510cb16b1ae3f/docker-java-core/src/main/java/com/github/dockerjava/core/dockerfile/Dockerfile.java#L186).

**Comment by viatra2** [comment]
More specifically, [this recursion](https://github.com/docker-java/docker-java/blob/e1586ccb94140d6c3c2cfd28d50510cb16b1ae3f/docker-java-core/src/main/java/com/github/dockerjava/core/dockerfile/Dockerfile.java#L186) always goes deep into the folder even if the current `directory` is ignored AND .dockerignore does not have exclusions for any paths below it (AFAIU this exclusion would be something like `!directory/evenDeeper/doNotIgnoreMe*` entry in the .dockerignore file)

**Comment by borlander** [comment]
AFAIU the Docker CLI code also walks the whole folder structure without further optimizations [here](https://github.com/docker/cli/blob/8f14db8df2fc9ee5f015df2fa3a3fd1e1d8e0b71/cli/command/image/build/context.go#L50)

**Comment by swiedenfeld** [comment]
@borlander Docker CLI is written in Go, and this is written Java. I don't think they can be compared.  Looking at the code, ``ScannedResult``  uses ``java.io``, and it might be a bottleneck traversing the large directory. Maybe ``java.nio`` could improve processing?

**Comment by swiedenfeld** [comment]
I did a bit of research on this. 

At first glance, the idea of @viatra2 in https://github.com/docker-java/docker-java/issues/1409#issuecomment-640615130 sounds correct and promising.  

Following, I implemented a file walk with ``java.nio.file.Files.walkFileTree``, but it did not traverse the tree considerably faster. 

Then, I did a short circuit that prevents the ``node_modules`` to be traversed further if it is ignored. But the exception rules ``!`` rules bit me. There is a lot of combinations of exclude and include lines possible, plus their order in ``.dockerignore`` matters. It may thus be inevitable to visit each file and check their effective inclusion/exclusion.

Nevertheless, one possible improvement I found is #1411.
